core/src/main/java/org/apache/calcite/sql/SqlSelect.java

```diff
@@ -288,4 +297,146 @@ public boolean hasWhere() {
     return getModifierNode(targetKeyWord) != null;
   }
+
+  /**
+   * Rewrite a select statement with a where condition if the where
+   * condition depends on the columns in the select list. This writes
+   * the code to use two select statements.
+   *
+   * For example <code>Select A + 1 as X from table where X &gt; 3</code>
+   * becomes
+   * Select X
+   *  FROM
+   *    (Select A + 1 as X)
+   * where X &gt; 3
+   *
+   * We only do this replacement if we detect that the where may reference an alias generated
+   * by the select statement. This approach simplifies code but is not required for correctness.
+   * If this fails then the alias will be inlined as a fallback in validation.
+   *
+   * @return The new top level select
+   */
+  public SqlSelect rewriteSqlSelectIfWhereAlias() {
+    // Right now to make sure everything executes in order we require
+    // a select statement only contain a where
+    if (!keywordList.isEmpty() || where == null || groupBy != null
+        || having != null || qualify != null || !windowDecls.isEmpty()
+        || orderBy != null || offset != null || fetch != null) {
+      // Only perform the rewrite if we just have a where.
+      return this;
+    }
+    // Create a set of identifiers that we are keeping. If we have an alias
+    // that matches an identifier then we cannot safely convert this query into
+    // two selects
+    Set<ImmutableList<String>> keptIdentifiers = new HashSet<>();
+    // Create a set of Alias names that could be referenced by a where
+    Set<ImmutableList<String>> possibleAliases = new HashSet<>();
+    // Iterate through the select statements and check for any aliases that
+    // may be used in the where.
+    for (int i = 0; i < selectList.size(); i++) {
+      SqlNode selectVal = selectList.get(i);
+      if (selectVal instanceof SqlBasicCall && selectVal.getKind() == SqlKind.AS) {
+        SqlBasicCall innerAlias = (SqlBasicCall) selectVal;
+        // Check if the alias is unnecessary. If so skip the transformation.
+        List<SqlNode> operands = innerAlias.getOperandList();
+        SqlIdentifier aliasIdentifier = (SqlIdentifier) operands.get(operands.size() - 1);
+        if (keptIdentifiers.contains(aliasIdentifier.names)) {
+          // We cannot safely perform this transformation as we may have a name conflict.
+          // We don't need to populate all identifiers first because this would fail to
+          // validate.
+          return this;
+        }
+        if (operands.size() == 2 && (operands.get(0) instanceof SqlIdentifier
+            || operands.get(0) instanceof SqlLiteral)) {
+          // No need to replace if the alias is a simple column
+          // rename or a simple name for a literal.
+          continue;
+        }
+        possibleAliases.add(aliasIdentifier.names);
+      } else if (selectVal instanceof SqlIdentifier) {
+        keptIdentifiers.add(((SqlIdentifier) selectVal).names);
```

Comment:
I don't think this resolves the issue. Assume table1 has columns A, B, C
User query is: 

`SELECT (A+B+C) as A from table1 where A > 3`

In this case, keptIdentifiers will remain empty, and possibleAliases will consist of one value, "A".

Since we have a possible alias and the alias is used in the where clause we will proceed. Inner Nodes will become `(A+B+C) as A, *` (due to the addition of * on line 430), which will eventually lead to the same issue as above:

`Select A from (Select (A+B+C) as A, A, B, C from table1) where A > 3`

Reply:
I realize now that we don't know the names of the fields that are input to this select without being slightly further in Validation.

I may have lied to you about where this needs to take place, if this needs to take place later during validation, you may need to handle scoping BS (which is why I thought it would be better to do it here.)